src/video_core/page_manager.cpp

```diff
@@ -256,21 +274,23 @@ struct PageManager::Impl {
         }
 
         size_t base_page = (base_addr >> PAGE_BITS);
+        ASSERT(base_page % PAGES_PER_LOCK == 0);
+        std::scoped_lock lk(locks[base_page / PAGES_PER_LOCK]);
         auto perms = cached_pages[base_page + start_range.first].Perm();
         u64 range_begin = 0;
         u64 range_bytes = 0;
+        u64 potential_range_bytes = 0;
 
         const auto release_pending = [&] {
             if (range_bytes > 0) {
                 RENDERER_TRACE;
                 // Perform pending (un)protect action
                 Protect((range_begin << PAGE_BITS), range_bytes, perms);
                 range_bytes = 0;
+                potential_range_bytes = 0;
             }
         };
 
-        std::scoped_lock lk(lock);
-
         // Iterate pages
         for (size_t page = start_range.first; page < end_range.second; ++page) {
```

Comment:
start_range  and end_range appear to be used as a general start and end boundary of the update and are computed twice (one time in region manager and one here). I think in region manager you could use iterators on the mask to get start and end, and then pass the start and end pages in this function to avoid recomputing them twice